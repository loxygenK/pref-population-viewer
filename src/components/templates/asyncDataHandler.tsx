import React from "react";
import Skeleton, { SkeletonProps } from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";

import styles from "./asyncDataHandler.module.scss";

export interface AsyncDataHandlerProps<T> {
  data: T | undefined;
  error: unknown | undefined;
  children: (settledData: T) => React.ReactElement;
  skeletonProps?: SkeletonProps;
}
export const AsyncDataHandler = <T,>({
  data,
  error,
  children,
  skeletonProps,
}: AsyncDataHandlerProps<T>): React.ReactElement => {
  if (error !== undefined) {
    const description = tryStringify(error) ?? "不明なエラーです";

    return (
      <div>
        <h3 className={styles.error_heading}>エラーが発生しました</h3>
        <p>{description}</p>
      </div>
    );
  }

  return data !== undefined ? children(data) : <Skeleton {...skeletonProps} />;
};

const tryStringify = (value: unknown): string | undefined => {
  if (value == null) {
    return undefined;
  }

  if (typeof value !== "object") {
    return String(value);
  }

  if (value instanceof Error) {
    return value.message;
  }

  if (Object.keys(value).includes("toString")) {
    return value.toString();
  }

  return undefined;
};
