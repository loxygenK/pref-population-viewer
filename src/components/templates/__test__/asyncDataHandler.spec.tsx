import renderer from "react-test-renderer";
import { AsyncDataHandler } from "../asyncDataHandler";

describe("Async data handler", () => {
  it("renders the children with resolved data when data is available", async () => {
    const tree = renderer
      .create(
        <AsyncDataHandler data="Some data" error={undefined}>
          {(data) => <p>{data}</p>}
        </AsyncDataHandler>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders the error screen when error is provided", async () => {
    const tree = renderer
      .create(
        <AsyncDataHandler data={undefined} error={new Error("It failed")}>
          {(_) => {
            fail("This must not be rendered!");
          }}
        </AsyncDataHandler>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
