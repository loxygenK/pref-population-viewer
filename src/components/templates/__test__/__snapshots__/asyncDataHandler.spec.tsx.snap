// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async data handler renders the children with resolved data when data is available 1`] = `
<p>
  Some data
</p>
`;

exports[`Async data handler renders the error screen when error is provided 1`] = `
<div>
  <h3
    className="error_heading"
  >
    エラーが発生しました
  </h3>
  <p>
    It failed
  </p>
</div>
`;
